File: heapSort.js
let arr = [6, 13, 10, 4, 1, 5, 2, 8, 14, 9, 11, 7, 3, 15, 12];

function buildMaxHeap() {
  heapSize = arr.length - 1;
  //Math.floor(heapSize / 2)其實就是parent node,所以其實就是找最後一個parent node開始 一路往前面去整理
  for (let i = Math.floor(heapSize / 2); i >= 0; i--) {
    maxheapify(i);
  }
}

function maxheapify(i) {
  let largest;
  let l = i * 2 + 1;
  let r = i * 2 + 2;
  if (l <= heapSize && arr[l] > arr[i]) {
    largest = l;
  } else {
    largest = i;
  }
  if (r <= heapSize && arr[r] > arr[largest]) {
    largest = r;
  }
  if (i != largest) {
    let tmp = arr[largest];
    arr[largest] = arr[i];
    arr[i] = tmp;
    maxheapify(largest);
  }
}

function heapSort() {
  buildMaxHeap();
  for (let i = arr.length - 1; i >= 0; i--) {
    let temp = arr[0];
    arr[0] = arr[i];
    arr[i] = temp;
    heapSize--; //在maxheapify裡面檢查大小(largest)的時候會排除目前換到最後一個(最後一個會是目前循環的最大值)
    maxheapify(0); //再重頭執行
  }
  return arr;
}

heapSort();
console.log(arr);

File: heapSort.py
def heapsort():
    buidmaxheap()
    for i in range(len(arr)-1,-1,-1):
        arr[0] ,arr[i] =arr[i],arr[0]
        maxheapify(0,i-1) #每循環一次,因為最後一個元素不再進行maxheapify(最大值),所以每次長度都固定少一

def buidmaxheap():
    for i in range((len(arr)-1)//2,0-1,-1):
        maxheapify(i,(len(arr)-1)) #len(arr)-1純粹是讓maxheapify的 l 跟r限定在list內


def maxheapify(n,arr_langh):
    largest = None
    l = n*2 +1 
    r = n*2 +2
    if (l <= arr_langh) and (arr[l] > arr[n]):
        largest = l
    else:
        largest = n
    if (r <= arr_langh) and arr[r] > arr[largest]:
        largest = r
    if n != largest:
        tmp = arr[n]
        arr[n] = arr[largest]
        arr[largest] = tmp
        maxheapify(largest,arr_langh)


if __name__ == '__main__':
    arr = [15,669,45,51,848,6,21,48,954,877,32,54,65,3345,31,248,35,8,9999]
    heapsort()
    print(arr)
